utils/jjb_client.py

```diff
@@ -0,0 +1,184 @@
+
+            fetch_state = 'desired' if compare else 'current'
+            output_dir = path.join(io_dir, 'jjb', fetch_state, name)
+            args = ['--conf', ini_path,
+                    'test', config_path,
+                    '-o', output_dir,
+                    '--config-xml']
+            self.execute(args)
+
+        if compare:
+            self.print_diffs(io_dir)
+
+    def print_diffs(self, io_dir):
+        current_path = path.join(io_dir, 'jjb', 'current')
+        current_files = self.get_files(current_path)
+        desired_path = path.join(io_dir, 'jjb', 'desired')
+        desired_files = self.get_files(desired_path)
+
+        create = self.compare_files(desired_files, current_files)
+        delete = self.compare_files(current_files, desired_files)
+        common = self.compare_files(desired_files, current_files, in_op=True)
+
+        self.print_diff(create, desired_path, 'create')
+        self.print_diff(delete, current_path, 'delete')
+        self.print_diff(common, desired_path, 'update')
+
+    def print_diff(self, files, replace_path, action):
+        for f in files:
+            if action == 'update':
+                ft = self.toggle_cd(f)
+                equal = filecmp.cmp(f, ft)
+                if equal:
+                    continue
+
+            instance, item, _ = f.replace(replace_path + '/', '').split('/')
+            item_type = et.parse(f).getroot().tag
+            item_type = item_type.replace('hudson.model.ListView', 'view')
+            item_type = item_type.replace('project', 'job')
+            logging.info([action, item_type, instance, item])
+
+    def compare_files(self, from_files, subtract_files, in_op=False):
+        return [f for f in from_files
+                if (self.toggle_cd(f) in subtract_files) is in_op]
+
+    def get_files(self, search_path):
+        return [path.join(root, f)
+                for root, _, files in os.walk(search_path)
+                for f in files]
+
+    def toggle_cd(self, file_name):
+        if 'desired' in file_name:
+            return file_name.replace('desired', 'current')
+        else:
+            return file_name.replace('current', 'desired')
+
+    def update(self):
+        for name, wd in self.working_dirs.items():
+            ini_path = '{}/{}.ini'.format(wd, name)
+            config_path = '{}/config.yaml'.format(wd)
+
+            args = ['--conf', ini_path, 'update', config_path, '--delete-old']
+            self.execute(args)
+
+    def execute(self, args):
+        os.environ['PYTHONHTTPSVERIFY'] = self.python_https_verify
+
+        jjb = JenkinsJobs(args)
+        with self.toggle_logger():
+            jjb.execute()
+
+    def modify_logger(self):
+        formatter = logging.Formatter('%(levelname)s: %(message)s')
+        logger = logging.getLogger()
+        logger.handlers[0].setFormatter(formatter)
+        self.default_logging = logger.level
+
+    @contextmanager
+    def toggle_logger(self):
```

Comment:
:heart:

Reply:
very clever way to solve the JJB logging problem!

Reply:
:heartbeat: